File: 11505/main.py
import sys
from math import *
input = sys.stdin.readline
n, m, k = map(int, input().split())
h = ceil(log2(n)) + 1
data = [1] * (n + 1)
tree = [1] * (1 << h)


def update(node, start, end, idx, val):
    if start > idx or idx > end:
        return
    if start == end:
        tree[node] = val
        return
    mid = (start + end) // 2
    update(node * 2, start, mid, idx, val)
    update(node * 2 + 1, mid + 1, end, idx, val)
    tree[node] = (tree[node * 2] * tree[node * 2 + 1]) % 1000000007


def query(node, start, end, left, right):
    if start > right or left > end:
        return 1
    if left <= start and end <= right:
        return tree[node]
    mid = (start + end) // 2
    return (query(node * 2, start, mid, left, right) * query(node * 2 + 1, mid + 1, end, left, right)) % 1000000007


for i in range(n):
    x = int(input())
    update(1, 0, (1 << (h - 1)) - 1, i, x)

for i in range(m + k):
    a, b, c = map(int, input().split())
    if a == 1:
        update(1, 0, (1 << (h - 1)) - 1, b - 1, c)
    elif a == 2:
        print(query(1, 0, (1 << (h - 1)) - 1, b - 1, c - 1))

File: 11505/cpp/main.cpp
#include <iostream>
#include <algorithm>
#include <queue>
#include <stack>
#include <vector>
#include <cstdio>
#include <math.h>

#define MOD 1000000007

using namespace std;
typedef long long ll;
vector<ll> inits;

void init(vector<ll> &tree, int node, int start, int end, int idx, int val){
    if (start > idx || end < idx)
        return;
    if (start == end){
        tree[node] = val % MOD;
        return;
    }
    int mid = (start + end) / 2;
    init(tree, node * 2, start, mid, idx, val);
    init(tree, node * 2 + 1, mid + 1, end, idx, val);
    tree[node] = (tree[node * 2] * tree[node * 2 + 1]) % MOD;
}

ll query(vector<ll> &tree, int node, int start, int end, int left, int right){
    if (left > end || start  > right){
        return 1;
    }
    if (left <= start && end <= right){
        return tree[node];
    }
    int mid = (start + end) / 2;
    return (query(tree, node * 2, start, mid, left, right) * query(tree, node * 2 + 1, mid + 1, end, left, right)) % MOD;
}

int main() {
    int n, m, k;
    scanf("%d %d %d", &n, &m, &k);
    int h = (int)ceil(log2(n)) + 1;
    int size = (1 << h);
    int start = 0;
    int end = (1 << (h - 1)) - 1;
    vector<ll> tree(size);
    for(int i = 0; i < n; i ++){
        int val;
        scanf("%d", &val);
        init(tree, 1, start, end, i, val);
    }
    for(int i = 0; i < m + k; i ++){
        int a, b, c;
        scanf("%d %d %d", &a, &b, &c);
        if (a == 1){
            init(tree, 1, start, end, b - 1, c);
        }
        else if (a == 2){
            printf("%lld\n", query(tree, 1, start, end, b - 1, c - 1));
        }
    }
}
